File: Python3/src/0001 ~ 0100/0004 - Median of Two Sorted Arrays.py
# 链接：https://leetcode.com/problems/median-of-two-sorted-arrays/
# 题意：给定两个有序的数组，返回它们的中位数？

# 输入： nums1 = [1,3], nums2 = [2]
# 输出： 2.00000
# 解释： 合并后的数组为 [1,2,3], 中位数是 2

# 输入： nums1 = [1,2], nums2 = [3,4]
# 输出： 2.50000
# 解释： 合并后的数组为 [1,2,3,4], 中位数是 (2 + 3) / 2 = 2.5

# 输入： nums1 = [0,0], nums2 = [0,0]
# 输出： 0.00000
# 解释： 合并后的数组为 [0,0,0,0], 中位数是 (0 + 0) / 2 = 0

# 输入： nums1 = [], nums2 = [1]
# 输出： 1.00000
# 解释： 合并后的数组为 [1], 中位数是 1

# 思路： 二分
#
#       这题堪称劝退题，第一次做根本想不到 O(logn) 如何去解决，虽然再做时已经忘了具体解法，
#       但解决思路的关键点仍旧记得：充分利用已知信息——中位数，来确定数组的分割位置即可。
#
#       两个数组分别分割如下 nums1[:i], nums1[i:] 和 nums2[:j], nums2[j:] ，
#       我们假设中位数从 nums1[i - 1], nums1[i], nums2[j - 1], nums2[j] 中产生，
#       那么必有必有如下大小关系：nums1[i - 1] <= nums2[j] && nums2[j - 1] <= nums1[i]
#       且有如下相等关系：i + j == (m + n) / 2
#
#       1. 如果 m + n 是奇数，则有 i + j = (m - i) + (n - j) - 1 ，
#           即中位数为 min(nums1[i], nums2[j])
#       2. 如果 m + n 是偶数，则有 i + j = (m - i) + (n - j) ，
#           即中位数为 (max(nums1[i - 1], nums2[j - 1]) + min(nums1[i], nums2[j])) / 2
#
#       现在我们再考虑分割点不恰好再中位数时的情况，即前面提到的大小关系和相等关系有一个不满足，
#       那我们直接保证相等关系满足，这样就只有大小关系不满足着一种可能。
#       假设当前 nums1 被分成 nums1[:i] 和 nums1[i:] ，那么为了保证相等关系满足，
#       nums2 被分成 nums2[:j] 和 nums2[j:]，分割点 j = (m + n) / 2 - i 。
#
#       此时我们判断分割点附近四个数的大小，分别处理即可：
#       1. nums1[i - 1] <= nums2[j] && nums2[j - 1] <= nums1[i]:
#           则当前分割点以满足大小关系，根据前面提到的方式计算出中位数即可
#       2. nums1[i - 1] <= nums2[j] && nums2[j - 1] > nums1[i]:
#           则此时有： nums1[i - 1]  <= nums1[i] < nums2[j - 1] <= nums2[j] ，
#           按照正确的分割点，应该要保证 nums2[j - 1] <= nums1[i] ，
#           说明 nums1[i] 偏小了，正确分割点在 (i, m) 内
#       3. nums1[i - 1] > nums2[j] && nums2[j - 1] <= nums1[i]:
#           则此时有： nums2[j - 1] <= nums2[j] < nums1[i - 1] <= nums1[i] ，
#           按照正确的分割点，应该要保证 nums1[i - 1] <= nums2[i] ，
#           说明 nums1[i - 1] 偏大了，正确分割点在 [0, i) 内
#       4. nums1[i - 1] > nums2[j] && nums2[j - 1] > nums1[i]:
#           则此时有： nums1[i - 1] <= nums1[i] < nums2[j - 1] <= nums2[j] < nums1[i - 1]
#           由于存在严格小于，所以这种情况不存在，可以直接排除
#
#       时间复杂度： O(logn)
#       空间复杂度： O(1)

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        m, n = len(nums1), len(nums2)
        # 保证 nums1 长度更短，同时可以更好处理边界情况
        if m > n:
            nums1, nums2 = nums2, nums1
            m, n = n, m

        half = (m + n) // 2
        # 二分 nums1 的区间 [0, m)
        l, r = 0, m
        while l < r:
            # nums1 在 [l, r) 内取中间的位置为分割点
            i = (l + r) // 2
            # 计算出 nums2 的分割点
            j = half - i
            # 判断此时分割点数字的大小关系
            if j > 0 and nums2[j - 1] > nums1[i]:
                # 情况 2 ，正确分割点在 (i, m) 内
                l = i + 1
            else:
                # 情况 3 ，正确分割点在 [0, i) 内
                # 情况 1 ，即使是一个正确的分割点，但是区间大小没有收缩到 1 ，需要继续处理
                #   这样最终 nums1 正确的分割点就是 l
                r = i

            # 情况 4 不存在

        # 目前已找到正确分割点 l
        i, j = l, half - l
        # 找到右边部分两个数的较小值，注意处理好边界情况
        if i == m:
            right_min = nums2[j]
        elif j == n:
            right_min = nums1[i]
        else:
            right_min = min(nums1[i], nums2[j])

        # 如果是奇数，则中位数就是右边部分两个数的较小值
        if (m + n) & 1:
            return right_min

        # 找到左边部分两个数的较大值，注意处理好边界情况
        if i == 0:
            left_max = nums2[j - 1]
        elif j == 0:
            left_max = nums1[i - 1]
        else:
            left_max = max(nums1[i - 1], nums2[j - 1])

        # 偶数，中位数就是 (左边部分两个数的较大值 + 右边部分两个数的较小值) / 2
        return (left_max + right_min) / 2

File: Go/src/0001 ~ 0100/0004 - Median of Two Sorted Arrays.go
// 链接：https://leetcode.com/problems/median-of-two-sorted-arrays/
// 题意：给定两个有序数组 a 和 b，长度分别为 m 和 n，使用时间复杂度不超过 O(log(m + n)) 的算法求出两个数组合并后的中位数
// 思路：二分

// 一看时间复杂度就知道要二分，但是想了半天，一直认为要对两个数组都二分，导致很复杂，思路也比较混乱
// 看了题解一半就明白如何去做了，还是要会分析，利用已知的信息降低复杂度

// 假设两个数组的分割点分别是 i, j，即分为 a[0]...a[i - 1] | a[i]...a[m - 1] 和 b[0]...b[j - 1] | b[j]...b[n - 1]
// 且中点值从 a[i - 1]、a[i]、b[j - 1]、b[j] 中产生
// 设 i + j = lenLeft，(m - i) + (n - j) = lenRight，m <= n
// 1. 若 m + n 是奇数
// 		则必有：
// 		(1) lenLeft == lenRight - 1
// 		(2) a[i - 1] <= b[j] && b[j - 1] <= a[i]
//		且中点值为：min(a[i], b[j])
// 2. 若 m + n 是偶数
// 		则必有：
// 		(1) lenLeft == lenRight
// 		(2) a[i - 1] <= b[j] && b[j - 1] <= a[i]
//		且中点值为：float64(max(a[i - 1], b[j - 1]) + min(a[i], b[j])) / 2
// 综上，我们可以统一找到分割点，然后根据 m + n 的奇偶性返回中点值
// 条件可以统一为：
// (1) lenLeft = (lenLeft + lenRight) >> 1 = (m + n) >> 1
// (2) a[i - 1] <= b[j] && b[j - 1] <= a[i]
// 二分 数组a，得到 i，从而为了满足 条件(1) 可以计算出 j = half - i，然后判断 条件(2) 是否满足即可
// 若 条件(2) 满足，则二分结束，返回相应的中点值即可
// 若 条件(2) 不满足：
//		若 a[i - 1] > b[j]，则 a 数组的分割点必定可以左移（r = i - 1），即 i = 左边区间的中点
// 		若 b[j - 1] > a[i]，则 a 数组的分割点必定可以右移（l = i + 1），即 i = 右边区间的中点
//		【注意】不可能出现 a[i - 1] > b[j] && b[j - 1] > a[i]
// 同时，可以适当，将上面三种情况融合成两种，即调整左移的端点 r = i，即可满足 条件(2) 也并入，最终 l 即 i 的分割点
// 【注意】同时还要边界情况，防止越界
// 实际上由于 m < n，所以
// 		j == 0 时必定有 i >= m
// 		j == n 时必定有 i == 0 && m == n
func findMedianSortedArrays(a []int, b []int) float64 {
	m, n := len(a), len(b)
	// 让 a 的长度更小，使算法复杂度为：O(log(min(m, n)))
	if m > n {
		a, b = b, a
		m, n = n, m
	}
	half := (m + n) >> 1 // 由于 m <= n，则必定有：m <= half && lenLeft ==  half

	l, r := 0, m
	// 二分
	for ; l < r;  {
		i := (l + r) >> 1
		j := half - i
		// 【注意】不可能出现 a[i - 1] > b[j] && b[j - 1] > a[i]
		if j > 0 && b[j - 1] > a[i] {
			// 此处要取 i + 1，因为本次判定 i 必定不能为分界线，所以要从右区间要从 i + 1 开始
			l = i + 1 //若 b[j - 1] > a[i]，则 a 数组的分隔点必定可以右移，即 i = 右边区间的中点
		} else {
			r = i // 若 a[i - 1] > b[j]，则 a 数组的分隔点必定可以左移，即 i = 左边区间的中点
		}
	}

	i := l
	j := half - i
	var rightMin int
	// 处理好边界情况，防止越界
	if i == m {
		rightMin = b[j]
	} else if j == n {
		rightMin = a[i]
	} else {
		rightMin = min(a[i], b[j])
	}
	if ((m + n) & 1) == 1 {
		// 长度为奇数
		return float64(rightMin)
	}

	var leftMax int
	if i == 0 {
		leftMax = b[j - 1]
	} else if j == 0 {
		leftMax = a[i - 1]
	} else {
		leftMax = max(a[i - 1], b[j - 1])
	}
	// 长度为偶数
	return float64(leftMax + rightMin) / 2.0
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}